### ISSUE TITLE
Tasks order and scheduling

### ISSUE THREAD

**Issue by jmichel-otb** [dask/dask#5584]
First of all I apologies because this is more a question than an issue or feature request (though it might turn to one of those in the end).

I am having a really hard time to figure out how to influence task scheduling to get it the way I want it.

Here is my setup. I have two set of tasks, say A and B. Each set has a few thousands tasks in it. A tasks depends on nothing. Each B task depends on a few A tasks output (5 in average).

In my code I delay all A tasks, then I delay B tasks using delayed A objects to model dependencies.

Then I call `client.compute(B_delayed)` and I retrieve results of B on master node on the flow, using `as_completed` (client is from `dask.distributed`).

I would expect the scheduler to figure out that each B task has very few dependencies to A tasks, and therefore to try to complete entire B tasks to be able to clear ressources for A tasks that are no longer needed.

What happens is quite the opposite : the scheduler prioritize A tasks, trying to complete almost all of them while very scarcely computing B tasks (even when a lot of them become ready to run).

This is problematic for us:
- On the flow retrieval is less efficient if all B tasks come at the end (it involves IO operations)
- In some cases, the amount of memory on workers is insufficient to store all A results. This memory could be easily cleared, if only the scheduler decided to complete the corresponding B tasks.

What I tried so far:
- Also submit A tasks with `client.compute()`, with a lower priority (did not change the behavior)
- Submit B tasks in batch with decreasing priority (it helps a bit, but scheduler still wants to complete as many A tasks as possible).
- Mix of both : no noticeable progress

**Comment by mrocklin** [comment]
This page may interest you:
https://distributed.dask.org/en/latest/priority.html

In general though if you are submitting a single graph with `dask.compute`
then I would expect Dask to choose a prioritization that minimized memory
footprint.  It's heuristics are usually pretty good at this.  If you have
task graphs that are only a few hundred nodes then you can visualize the
prioritization by calling `dask.visualize(..., color="order")`

**Comment by jmichel-otb** [comment]
Thanks for your answer. I submit work with a single compute on a list of delayed B. One thing to note is that A code calls C/C++ code under the hood, so dask may not be aware of all the memory used. 

I have been able to visualize the graph on a subset my problem, and I can confirm that dask mostly prioritize A tasks over B tasks (I can share the graph plot if that helps).

**Comment by mrocklin** [comment]
Sure

**Comment by jmichel-otb** [comment]
So this is a code sample simulating what I am actually doing:

```python
import xarray as xr
import numpy as np
import dask
from dask.distributed import as_completed
from time import sleep

task_length = 2
a_size = 1000

# First task in graph
def A(value):
    coord = np.array(range(0,a_size))
    buffer = np.ndarray(shape=(a_size,a_size),dtype=float)
    buffer.fill(value)

    output = xr.Dataset({'a' : (['c1','c2'] , buffer),
                         'b' : (['c1','c2'] , buffer),
                         'c' : (['c1','c2'] , buffer)},
                        coords = {'c1' : coord, 'c2' : coord})
    sleep(task_length)
    return output

b_size = 500

# second task in graph
def B(inputs,value):
    coord = np.array(range(0,b_size))
    buffer = np.ndarray(shape=(b_size,b_size),dtype=float)
    buffer.fill(value)

    output = xr.Dataset({'a' : (['c1','c2'] , buffer),
                         'b' : (['c1','c2'] , buffer),
                         'c' : (['c1','c2'] , buffer)},
                        coords = {'c1' : coord, 'c2' : coord})

    sleep(task_length)
    return output

a_delayed = []

nb_tasks = 20
nb_deps = 3

# Submit all A tasks as delayed
for i in range(0,nb_tasks):
    a_delayed.append(dask.delayed(A)(i))

# Submit all B tasks as delayed, each B task depend on 3 A tasks
b_delayed = []
for i in range(0,nb_tasks):
    b_delayed.append(dask.delayed(B)(a_delayed[i:min(i+nb_deps,nb_tasks)],i))

# Display graph
dask.visualize(b_delayed,filename='graph.png',color='order', cmap='autumn')
```

I tried to return `xarray` with approximate size matching what is done in the actual code. I am not sure this information is used by dask but I also use this code to simulate actual execution on a distributed cluster.

You can find attached the displayed graph.

![graph](https://user-images.githubusercontent.com/10595679/69046277-de0aa380-09f8-11ea-8c58-b36270a78cc7.png)
You can see that task 39 is the last one that will be computed, which means dask will have to keep 19, 26 and 28 until the end. With this setup I would expect an order such as 

A20->B20->A19->B19->A18->B18 (at this point A20 can be released)
->A17 -> B17 (at this point A19 can be released) 
and so on.

Now if instead of 2 x 20 tasks we have 2 x 10 000 this effect is emphasised and I end up with workers memory problem.

**Comment by guillaumeeb** [comment]
Just chiming in to ask if this may be related to https://github.com/dask/distributed/issues/2602? A simpler vision of the problem?

**Comment by TomAugspurger** [comment]
It's possibly similar, though there are a few issues at play. It's important to separate

1. `dask.order.order` getting a bad linear ordering of tasks
2. The distributed scheduler not understanding later dependency structure when assigning a batch of tasks to idle workers (see https://github.com/dask/distributed/issues/2602#issuecomment-518297064)

I'm not sure which one this is yet.

FYI @eriknw is looking into this set of issues for the next little while. Hopefully he'll have more luck at a solution than I did :)

**Comment by jmichel-otb** [comment]
So I have been trying to trick dask into consuming tasks the way I want, using my knowledge of how tasks are organized (which dask tries to discover by itself during graph optimization if I understand well).

I tried options listed on this documentation page (as pointed out by @mrocklin). If I read it correctly, `priority` and `fifo_timeout` should allow to control priority between several calls to `client.compute()`.

So I formed groups of B tasks which I know have a lot of A tasks in common, and submit each group with a new call to `client.compute()`, using fifo_timout='0ms' parameter (as in the doc) to enforce the first in first out rule even if calls are close in time, and also using `priority=` parameters with a decreasing priority from one group to the other.

I must say that neither option seems to have any impact at all on how the tasks are dispatched by the scheduler. I still see it consume most of A tasks over B tasks. Did I mis-use something ?

How is the scheduler using this priority parameter ? Is there something else it takes into account (task average duration, number of tasks, number of dependencies of task ...) that is taken into which would explain that A tasks always get priority ?
